### PR TITLE
Tokens Module

### PR DESCRIPTION
**Pull request by shaopengw**
Version 1 impl for tokens module #4

### CODE REVIEW

File: tokens/src/lib.rs [hunk 1]
```diff
@@ -1,21 +1,45 @@
...
+// #3295 https://github.com/paritytech/substrate/issues/3295
+use srml_system::{self as system, ensure_signed};
+
 use traits::MultiCurrency;
 
-pub trait Trait: system::Trait {
-	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+pub trait Trait: srml_system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as srml_system::Trait>::Event>;
+	type Balance: Parameter + Member + SimpleArithmetic + Codec + Default + Copy + MaybeSerializeDeserialize + Debug;
```
Comment (xlc) [inline tokens/src/lib.rs]:
```suggestion
	type Balance: Parameter + Member + SimpleArithmetic + Default + Copy + MaybeSerializeDeserialize;
```

`Parameter` already includes `Codec` and `Debug`.

File: tokens/src/lib.rs [hunk 2]
```diff
@@ -1,21 +1,45 @@
...
 decl_storage! {
 	trait Store for Module<T: Trait> as Tokens {
+		/// The total issuance of a token type;
+		pub TotalIssuance get(fn total_issuance) build(|config: &GenesisConfig<T>| {
+			let issuance = config.initial_balance * (config.endowed_accounts.len() as u32).into();
+			config.tokens.iter().map(|id| (id.clone(), issuance)).collect::<Vec<_>>()
+		}): map T::CurrencyId => T::Balance;
 
+		/// The balance of a token type under an account.
+		pub Balance get(fn balance): double_map T::CurrencyId, twox_128(T::AccountId) => T::Balance;
```
Comment (xlc) [inline tokens/src/lib.rs]:
```suggestion
		pub Balance get(fn balance): double_map T::CurrencyId, blake2_256(T::AccountId) => T::Balance;
```

blake2_256 to be safe

File: tokens/src/lib.rs [hunk 3]
```diff
@@ -25,9 +49,88 @@ decl_module! {
...
+		);
+
+		<TotalIssuance<T>>::mutate(currency_id, |v| *v -= amount);
+		<Balance<T>>::mutate(currency_id, who, |v| *v -= amount);
+
+		Ok(())
+	}
+
+	fn slash(currency_id: &Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> Self::Balance {
+		let actual_amount = Self::balance(currency_id, who).max(amount);
```
Comment (xlc) [inline tokens/src/lib.rs]:
```suggestion
		let actual_amount = Self::balance(currency_id, who).min(amount);
```

File: traits/src/lib.rs [hunk 4]
```diff
@@ -1,3 +1,45 @@
...
+	fn total_inssuance(currency_id: &Self::CurrencyId) -> Self::Balance;
+
+	/// The combined balance of `who` under `currency_id`.
+	fn balance(currency_id: &Self::CurrencyId, who: &AccountId) -> Self::Balance;
+
+	// Public mutables
+
+	/// Transfer some amount from one account to another.
+	fn transfer(
+		currency_id: &Self::CurrencyId,
```
Comment (xlc) [inline traits/src/lib.rs]: `CurrencyId` already is `Copy` so you can just pass it by value

**Comment by xlc** [comment]
For next phrase:
- use `decl_error` and avoid string errors
- make tokens module works with `Currency` / `Balances` module

**Comment by shaopengw** [comment]
Will open other PRs for improvements.

As the ongoing "imbalances" types development would change most function signatures, unit tests would be added in other PRs along with these change.

**Pull request merged**